### PR TITLE
Add Array{T,N}(nothing/missing, dims) convenience constructors

### PR DESCRIPTION
**Pull request by nalimilan**
These are shorthands for `fill!(Array{T,N}(uninitialized, dims), nothing/missing)`. In the future they could be optimized for `isbits` `Union` types in which nothing/missing is the first type of the `Union`, since such arrays are already filled with `nothing`/`missing`.

This implements solution 3. from https://github.com/JuliaLang/julia/issues/24939, which got broad support.

I haven't found a way to attach the new docstrings to the new methods rather than to the general docstring for the function/type, given that methods are generated using `@eval` with a loop. Anyway the docstrings describe both the `nothing` and the `missing` method, so that sounds unavoidable to limit duplication.

### CODE REVIEW

File: base/docs/basedocs.jl [hunk 1]
```diff
@@ -1312,6 +1312,27 @@ julia> Vector{Float64}(uninitialized, 3)
  6.90966e-310
  6.90966e-310
 ```
+
+    Vector{T}(nothing, m)
```
Comment (fredrikekre) [inline base/docs/basedocs.jl]: I would prefer just splitting these into two docstrings and attach them to `Vector{T}(nothing, n)` and `Vector{T}(missing, n)` respectively. (And likewise for the `Matrix` and `Array` docstrings.)

Comment (nalimilan) [reply base/docs/basedocs.jl]: I wanted to do that too, but as I noted above I couldn't get it to work. Ideas?

Comment (fredrikekre) [reply base/docs/basedocs.jl]: I meant something like:
````julia
"""
    Vector{T}(uninitialized, n)

Construct an uninitialized [`Vector{T}`](@ref) of length `n`. See [`uninitialized`](@ref).

# Examples
```julia-repl
julia> Vector{Float64}(uninitialized, 3)
3-element Array{Float64,1}:
 6.90966e-310
 6.90966e-310
 6.90966e-310
```
"""
Vector{T}(uninitialized, n)

"""
    Vector{T}(nothing, m)

Construct a [`Vector{T}`](@ref) of length `m`, initialized with
[`nothing`](@ref) entries. Element type `T` must be able to hold
these values, i.e. `Void <: T`.

# Examples
```jldoctest
julia> Vector{Union{Void, String}}(nothing, 2)
2-element Array{Union{Void, String},1}:
 nothing
 nothing
```
"""
Vector{T}(nothing, n)

"""
    Vector{T}(missing, m)

Construct a [`Vector{T}`](@ref) of length `m`, initialized with
[`missing`](@ref) entries. Element type `T` must be able to hold
these values, i.e. `Void <: T`.

# Examples
```jldoctest
julia> Vector{Union{Void, String}}(missing, 2)
2-element Array{Union{Void, String},1}:
 missing
 missing
```
"""
Vector{T}(missing, n)
````

**Comment by nalimilan** [comment]
The new definitions cause weird redefinition warnings during bootstrap:
```
WARNING: Method definition (::Type{Array{T, N}})(Void, Integer...) where {T, N} in module Inference at array.jl:398 overwritten in module Base at array.jl:398.
WARNING: Method definition (::Type{Array{T, N}})(Void, Tuple{Vararg{Integer, N}}) where {T, N} in module Inference at array.jl:400 overwritten in module Base at array.jl:400.
WARNING: Method definition (::Type{Array{T, N} where N})(Void, Integer...) where {T, N} in module Inference at array.jl:403 overwritten in module Base at array.jl:403.
WARNING: Method definition (::Type{Array{T, N} where N})(Void, Tuple{Vararg{Integer, N}}) where {T, N} in module Inference at array.jl:405 overwritten in module Base at array.jl:405.
```

For some reason, they do not happen with the `::Missing` methods.

File: doc/src/stdlib/arrays.md [hunk 2]
```diff
@@ -7,15 +7,19 @@ Core.AbstractArray
 Base.AbstractVector
 Base.AbstractMatrix
 Core.Array
-Core.Array(::Any)
-Core.Array(::Any, ::Any)
+Core.Array(::Uninitialized, ::Any)
```
Comment (fredrikekre) [inline doc/src/stdlib/arrays.md]: Where these not included in the `Array` entry on the line above?

Comment (nalimilan) [reply doc/src/stdlib/arrays.md]: No, I got an error:
```
 !! No docs found for 'Core.Array(::Any, ::Any)'. [src/stdlib/arrays.md]
```
I thought there was a way to include all methods, but this one doesn't work at least.

Comment (fredrikekre) [reply doc/src/stdlib/arrays.md]: Right, I meant that you should remove `Core.Array(::Any)` and `Core.Array(::Any, ::Any)` and keep `Core.Array`. I *think* that `Core.Array` should catch everything associated with `Array`, but perhaps there was some reason for including them all earlier, perhaps to place the type docstring for `Array` at the top?

Comment (nalimilan) [reply doc/src/stdlib/arrays.md]: No, that only includes the type, not methods.

Comment (fredrikekre) [reply doc/src/stdlib/arrays.md]: Ok, good to know!

File: base/array.jl [hunk 3]
```diff
@@ -391,6 +391,22 @@ end
 one(x::AbstractMatrix{T}) where {T} = _one(one(T), x)
 oneunit(x::AbstractMatrix{T}) where {T} = _one(oneunit(T), x)
 
+# Constructors for nothing and missing
```
Comment (fredrikekre) [inline base/array.jl]: Unless I am missing something this whole block can be replaced by
```julia
# type and dimensionality specified
Array{T,N}(::Void, d...) where {T,N} = fill!(Array{T,N}(uninitialized, d...), nothing)
Array{T,N}(::Missing, d...) where {T,N} = fill!(Array{T,N}(uninitialized, d...), missing)
# type but not dimensionality specified
Array{T}(::Void, d...) where {T} = fill!(Array{T}(uninitialized, d...), nothing)
Array{T}(::Missing, d...) where {T} = fill!(Array{T}(uninitialized, d...), missing)
```
and just let the `Array{T[,N]}(::Uninitialized, d...)` constructor do the conversion etc of the input args.

Comment (nalimilan) [reply base/array.jl]: Yes, it could, but I followed the pattern used by `uninitialized` constructors in sysimg.jl. I'm not sure how much it matters, but I think the idea of the call to `convert` is to avoid compiling a specialized function for each integer type. So I'd rather keep that even if that's more verbose unless we're completely sure it's not needed.

**Comment by nalimilan** [comment]
I've found a way to avoid the "method redefined" warnings by defining the code in sysimg.jl together with existing `Array` constructors. I'll merge once CI passes unless somebody objects.

File: base/array.jl (hunk 3 above)
Comment (fredrikekre) [reply base/array.jl]: I don't understand, you introduce these constructors for all different types though? Wouldn't my suggestion just mean that we instead use another specialized `Array{T}(uninitialized, ...)` constructor?

Comment (nalimilan) [reply base/array.jl]: Yeah, but you still need to compile the method itself, the `Array{T}(uninitialized, ...)` method for `Integer`, and then the `Array{T}(uninitialized, ...)` method for `Int`. By doing the conversion directly, you can avoid one intermediate step. At least, that's how I understand it.

Comment (fredrikekre) [reply base/array.jl]: Yea, I just assumed that the `Array{T}(uninitialized, ...)` method for `Integer` was already compiled.

**Review by fredrikekre — APPROVED** [review]
LGTM, lets give @Sacha0 the opportunity to review since he is the one engineering these type of array constructors :)

**Comment by nalimilan** [comment]
AppVeyor failure on 64-bit is the infamous Bunch-Kaufman bug.

File: base/array.jl (hunk 3 above)
Comment (Sacha0) [reply base/array.jl]: I would err to code simplicity and clarity rather than worrying overmuch about compiling an additional method that likely will be compiled anyway :).

File: test/arrayops.jl [hunk 4]
```diff
@@ -2079,6 +2079,34 @@ end
...
+    @test all(x -> x isa U, b)
+    a = Array{Union{T, U}}(U(), 9,8,7,6,5,4,3,2,1)
+    b = Array{Union{T, U},9}(U(), 9,8,7,6,5,4,3,2,1)
+    @test size(a,1,1) == (9,9)
+    @test size(a,4) == 6
+    @test size(a,9,8,7,6,5,4,3,2,19,8,7,6,5,4,3,2,1) == (1,2,3,4,5,6,7,8,1,2,3,4,5,6,7,8,9)
+    @test size(a) == size(b)
+    @test all(x -> x isa U, a)
+    @test all(x -> x isa U, b)
+end
```
Comment (Sacha0) [inline test/arrayops.jl]: Why all the tests of `size`? They seem somewhat unrelated :).

File: base/sysimg.jl [hunk 5]
```diff
@@ -146,6 +146,21 @@ Matrix(::Uninitialized, m::Integer, n::Integer) = Matrix{Any}(uninitialized, Int
...
+            fill!(Array{T,N}(uninitialized, convert(Tuple{Vararg{Int}}, d)...), $U())
+        Array{T,N}(::$U, d::NTuple{N,Integer}) where {T,N} =
+            fill!(Array{T,N}(uninitialized, convert(Tuple{Vararg{Int}}, d)...), $U())
+        # type but not dimensionality specified
+        Array{T}(::$U, d::Vararg{Integer,N}) where {T,N} =
+            fill!(Array{T,N}(uninitialized, convert(Tuple{Vararg{Int}}, d)...), $U())
+        Array{T}(::$U, d::NTuple{N,Integer}) where {T,N} =
+            fill!(Array{T,N}(uninitialized, convert(Tuple{Vararg{Int}}, d)...), $U())
+    end
+end
```
Comment (Sacha0) [inline base/sysimg.jl]: I sympathize with `@fredrikekre` that simplifying / DRY'ing this block out a bit might be nice :).

File: test/arrayops.jl (hunk 4 above)
Comment (nalimilan) [reply test/arrayops.jl]: Indeed. I just copied the tests used for the standard constructors without noticing these extra lines.

File: base/sysimg.jl (hunk 5 above)
Comment (nalimilan) [reply base/sysimg.jl]: OK, I don't really care either way, so if everybody is against it... I've updated the PR to use the simpler form.
